### ISSUE TITLE
Heart rate from Wear app not updated

### ISSUE THREAD

**Issue by jimbolimbo3** [home-assistant/android#3100]
**Home Assistant Android version:**
2022.11.0
**Android version:**
12
**Phone model:**
OnePlus Nord 2
**Home Assistant version:**
2022.11.4
**Last working Home Assistant release (if known):**
2022.11.4
**Description of problem:**

The heart rate from my Samsung Galaxy Watch 4 is update every now and then, without any particular timing. (see screenshot from sensor). In the smart watch the heart rate is set to be read every 10 minutes, and I'm wearing it today from 8 am.

**Traceback (if applicable, to get the logs you may refer to: https://companion.home-assistant.io/docs/troubleshooting/faqs/#android-crash-logs):**

```

```

**Screenshot of problem:**
![Screenshot_2022-11-25-13-39-12-98_c3a231c25ed346e59462e84656a70e50](https://user-images.githubusercontent.com/2467359/203989067-bbddf045-f62e-4f34-849b-06888ee52c5f.jpg)
**Additional information:**

**Comment by jpelgrom** [comment]
Sensors update at ~15 minute intervals, although it will depend on Wear OS and your watch connectivity. Also see the [blue info box in the documentation](https://companion.home-assistant.io/docs/wear-os/sensors) about sensors on Wear OS. Any settings your watch may have for the sensor don't change this.

However, in your screenshot it looks like the sensor stopped updating completely since the 22nd. In the description you say it is updated "every now and then". To clarify, does it still update correctly, even if it is not immediately?

**Comment by damiaorocha** [comment]
I have the same problem. I also use a Samsung Galaxy Watch4. For example, the heart rate sensor was updated 7 hours ago, but the watch had a measurement 3 minutes ago. The other sensors are updating without problems. Battery and daily steps, for example, update whenever I turn on the watch screen.

**Comment by dshokouhi** [comment]
Can you guys confirm you are in the latest production version? We had a fix specifically for the sensor being stuck. Now keep in mind it doesn't matter how often other apps check the heart rate. What matters is that we get a valid reading when we request a sensor update. The app does indeed skip invalid readings. We may need to look at the logs but first confirm you are on the latest. On my pixel watch the heart rate updates as expected.

**Comment by damiaorocha** [comment]
I am using 2022.11.0-2948

**Comment by dshokouhi** [comment]
if you guys are able to get us the on device logs then we can look to see what happened.

**Comment by jimbolimbo3** [comment]
![Screenshot_2022-11-26-21-08-06-22_c3a231c25ed346e59462e84656a70e50.jpg](https://user-images.githubusercontent.com/2467359/204107194-65cf57f9-e857-4a06-a435-3f44fc0e9c9b.jpg)

Today I'm wearing the smartwatch from 16.00 to now, and it didn't send (or home assistant didn't get) any update.
Is there any method to get logs from the smartwatch, besides using adb?

**Comment by dshokouhi** [comment]
> Is there any method to get logs from the smartwatch, besides using adb?

adb logcat or android studio are the only method for retrieving the logs.

also home assistant does not get updates that other apps do. We request our own update and evaluate it and send an update based on that. The logs will show us if our requests are coming in or not.

**Comment by dshokouhi** [comment]
so the app will skip value when the system tells us the accuracy is not good enough. Meaning the accuracy reports there is no contact to the body or the sensor it not calibrated. We also skip values below 0 because those are invalid. The logs will tell us if we are getting updates or not which is important to know how to proceed here.

does the heart rate update when you guys open the app?

**Comment by VNRARA** [comment]
Mines is stuck at 0bpm on v2022.11.2

**Comment by dshokouhi** [comment]
all unfortunately without on device logs we cannot proceed here. We may need to add more logs but as of now we need to see what our code is doing on these devices.

**Comment by VNRARA** [comment]
Anything specific you need? https://developer.android.com/studio/command-line/logcat

**Comment by dshokouhi** [comment]
Yes we should be seeing the heart rate sensor listener being registered and unregistered when a sensor update is triggered like entering or leaving the app.

https://github.com/home-assistant/android/blob/master/wear/src/main/java/io/homeassistant/companion/android/sensors/HeartRateSensorManager.kt#L119

This line above in particular will tell us if we are getting a response

I can see in your logs here:

```
11-27 19:31:50.838 13574 13574 D HRSensor: Heart Rate sensor listener unregistered
```

Which suggests that we did get a response and it may or may not have been valid. `0` is a valid value technically speaking but we may be skipping it if the sensor reported the accuracy was not correct.

**Comment by dshokouhi** [comment]
So the issue users are seeing in this thread is that the updates stop coming in, thats why we want to see the logs in those case because if we see the listener being unregistered than that means the system is just not giving our app a good reading. I can add more logs so we can try to understand whats going on but given our current checks I am not sure it will be helpful. Its helpful enough to know we are getting a response but we need to discard certain values like bad readings which the sensor will report to us.

**Comment by damiaorocha** [comment]
Follow the logs. I inform you that the daily step sensor also stopped updating frequently.
[logs2.txt](https://github.com/home-assistant/android/files/10099420/logs2.txt)

**Comment by dshokouhi** [comment]
> Follow the logs.

Looks to be the same we are indeed getting responses from the heart rate sensor but for some reason they do not seem to be valid. I can submit a PR to add more debug logs so we can see what the values are to see what can be done.



> I inform you that the daily step sensor also stopped updating frequently.

lets not combine issues even though they may seem similar.  Please open a new issue and describe what you are experiencing. Just a FYI we have no control over how frequent the updates come in, we only report when we get data and each device reports data differently. Let us know if they stop completely or you are just experiencing a delay.

**Comment by dshokouhi** [comment]
ok can someoen please try the debug APK from:

https://github.com/home-assistant/android/actions/runs/3560485735

you will need to download the artifact, extract it and sideload the Wear APK. From there enable the heart rate sensor and let us see the logs so we can determine what is happening.

**Comment by VNRARA** [comment]
qq: How often is the heart rate sensor supposed to update (on the Samsung GW4 classic)?

**Comment by dshokouhi** [comment]
supposed to be every 15 minutes but can vary because if we get a bad reading we will discard it so could be an hour or 2 between updates. Really depends on what data we get. On my pixel watch sometimes I see a gap of about an hour. But that could be caused by a bad reading or the same state which results in no update.

**Comment by dshokouhi** [comment]
ok the debug logs have been merged, if possible please update to the latest beta (2966) and let us see the logs when you open the app with the sensor enabled. This should hopefully give us a clue as to what is going on.

**Comment by damiaorocha** [comment]
Here are the logs
[logs4.txt](https://github.com/home-assistant/android/files/10127714/logs4.txt)

**Comment by dshokouhi** [comment]
ok in this case it looks to be correct, the sensor reported to teh app that the accuracy is `NO_CONTACT`

```
2022-11-30 19:15:56.626  4379-4379  HRSensor                pid-4379                             D  HR event received with accuracy: no_contact and value: 0.0
```

https://developer.android.com/reference/android/hardware/SensorManager#SENSOR_STATUS_NO_CONTACT

@damiaorocha did you pull these logs on the charge or was the watch worn at the time?

**Comment by dshokouhi** [comment]
> these logs are from the moment I open the app. What defines which measurements are invalid?

This doesnt answer my question if the device was on teh charge or being worn :)



> What defines which measurements are invalid?

with each sensor event it reports the accuracy, we are printing the accuracy to see what is happening and in your case its not only reporting no contact but also a value of `0`



> I can see values ​​in the health app on the watch

For us this doesnt matter because we request an update from the heart rate sensor directly.

**Comment by damiaorocha** [comment]
It was in use

**Comment by damiaorocha** [comment]
Is there a possibility for the user to choose the precision?

**Comment by dshokouhi** [comment]
> Is there a possibility for the user to choose the precision?

no and how would that work if the sensor reports that it is not being worn :)

**Comment by damiaorocha** [comment]
Ok. Thanks

**Comment by dshokouhi** [comment]
We can only go by the data we get. The only thing I can think of in this case is to wait for like 5 events because maybe this device the first reading is not valid but a subsequent one is valid.

**Comment by damiaorocha** [comment]
ok

**Comment by dshokouhi** [comment]
would you be able to test a debug build if I submit an attempt to fix this? The debug app can be installed side by side the production app, all you need to do to avoid a conflict is to give the device a different name. Once the test is done you can uninstall the debug app. It may also require you to install the phone debug app, which the same rules apply with a different device name.

Although I can test the code myself I would prefer to know its working and the theory is correct :)

**Comment by damiaorocha** [comment]
I can try. But I never did.

**Comment by dshokouhi** [comment]
the hardest part is pushing the APK to the devices :) other than that its like having a second version of the app installed

**Comment by damiaorocha** [comment]
ok

**Comment by dshokouhi** [comment]
ok the debug build is ready

https://github.com/home-assistant/android/actions/runs/3588364792

you will need to download the artifact and extract it. From there you should see the Wear APK to install on the watch. When you login make sure to give a different device name so add a 1 or something to the end to make it unique. You may need to turn on your screen on the phone to get it to login or you may even need install the phone APK from the artifact. If you install the phone APK make sure to again give a different device name. First check to see if the sensor will eventually update its state by enabling it and putting the app in the background, opening the app again will trigger another update. Make sure to get the logs if you still experience an issue. 🤞  Try to make sure the logs are long enough to get all the events if its not updating. If you check the linked PR you can see examples of valid and invalid events.

**Comment by damiaorocha** [comment]
I installed the apk. And I could conclude that it still doesn't work well. Sometimes I open the app and it updates, sometimes it doesn't.

[logs6.txt](https://github.com/home-assistant/android/files/10147091/logs6.txt)

I also noticed problem in the values.

At 14:55 I requested a measurement of heartbeats according to logs:

2022-12-03 14:55:45.952 1759-2138 SHW - Hear...Repository pid-1759 I [startOnDemandMeasure][1670090145862, 73, FIND_HR]
2022-12-03 14:55:45.960 1759-21351 SHW - Hear...yViewModel pid-1759 I [startMeasureHeartRate] result [1670090145862, 73, FIND_HR]
2022-12-03 14:55:45.962 1759-1759 SHW - Hear...reActivity pid-1759 I [updateProgress] start animation, currentProgress:51
2022-12-03 14:55:45.967 1759-1759 SHW - Hear...reActivity pid-1759 I [measureFindHr]
2022-12-03 14:55:45.967 1759-1759 SHW - Hear...Repository pid-1759 I [stopOnDemandMeasure]<<<
2022-12-03 14:55:45.972 1759-1759 SHW - Hear...ckerForWhs pid-1759 I [stopOnDemandMeasure]
2022-12-03 14:55:45.980 1759-27462 SHW - Hear...Repository pid-1759 I [startOnDemandMeasure.awaitClose]
2022-12-03 14:55:45.980 1759-27462 SHW - Hear...Repository pid-1759 I [stopOnDemandMeasure]<<<
2022-12-03 14:55:45.989 1759-2138 SHW - HealthSensor pid-1759 I unregisterListener ~~listener 1 = 397d98d

The result of this measurement being 73.

further on I can see in the logs that the sensor has been updated with a value of 77. What is the reason for this difference?

2022-12-03 14:56:21.776 21221-21221 HRSensor                io....stant.companion.android.debug  D  HR event received with accuracy: high and value: 77.0 with event count: 4
....

2022-12-03 14:56:29.454 21221-24739 okhttp.OkHttpClient     io....stant.companion.android.debug  I  Content-Length: 690
2022-12-03 14:56:29.454 21221-24739 okhttp.OkHttpClient     io....stant.companion.android.debug  I  {"type":"update_sensor_states","data":[{"unique_id":"battery_power","state":0.0,"type":"sensor","icon":"mdi:battery-minus","attributes":{"current":-4.2E-5,"voltage":4.199}},{"unique_id":"heart_rate","state":"77","type":"sensor","icon":"mdi:heart-pulse","attributes":{"accuracy":"high"}},{"unique_id":"daily_floors","state":"","type":"sensor","icon":"mdi:stairs","attributes":{}},{"unique_id":"daily_distance","state":422.14062,"type":"sensor","icon":"mdi:map-marker-distance","attributes":{}},{"unique_id":"daily_calories","state":18.14154,"type":"sensor","icon":"mdi:fire","attributes":{}},{"unique_id":"daily_steps","state":555.0,"type":"sensor","icon":"mdi:shoe-print","attributes":{}}]}
2022-12-03 14:56:29.454 21221-24739 okhttp.OkHttpClient     io....stant.companion.android.debug  I  --> END POST (690-byte body)
2022-12-03 14:56:29.579 21221-24739 okhttp.OkHttpClient     io....stant.companion.android.debug  I  <-- 200 https://damiaohomeassistant.duckdns.org:8443/api/webhook/16c4df150bed37aeec7a6f66905ebc0e4ae623cffff0d9bdd0a0a4a29ec79e22 (124ms)
2022-12-03 14:56:29.580 21221-24739 okhttp.OkHttpClient     io....stant.companion.android.debug  I  server: nginx
2022-12-03 14:56:29.580 21221-24739 okhttp.OkHttpClient     io....stant.companion.android.debug  I  date: Sat, 03 Dec 2022 17:56:29 GMT
2022-12-03 14:56:29.581 21221-24739 okhttp.OkHttpClient     io....stant.companion.android.debug  I  content-type: application/json; charset=utf-8
2022-12-03 14:56:29.581 21221-24739 okhttp.OkHttpClient     io....stant.companion.android.debug  I  content-length: 212

[logs5.txt](https://github.com/home-assistant/android/files/10147095/logs5.txt)

**Comment by dshokouhi** [comment]
> Sometimes I open the app and it updates, sometimes it doesn't.

That is to be expected, don't expect updates each and everytime. The goal is to stop the graph from becoming flat for hours. Even in my own watch I see gaps up to 2 hours.

> I also noticed problem in the values.

> At 14:55 I requested a measurement of heartbeats according to logs:

```
2022-12-03 14:55:45.952 1759-2138 SHW - Hear...Repository pid-1759 I [startOnDemandMeasure][1670090145862, 73, FIND_HR]
2022-12-03 14:55:45.960 1759-21351 SHW - Hear...yViewModel pid-1759 I [startMeasureHeartRate] result [1670090145862, 73, FIND_HR]
2022-12-03 14:55:45.962 1759-1759 SHW - Hear...reActivity pid-1759 I [updateProgress] start animation, currentProgress:51
2022-12-03 14:55:45.967 1759-1759 SHW - Hear...reActivity pid-1759 I [measureFindHr]
2022-12-03 14:55:45.967 1759-1759 SHW - Hear...Repository pid-1759 I [stopOnDemandMeasure]<<<
2022-12-03 14:55:45.972 1759-1759 SHW - Hear...ckerForWhs pid-1759 I [stopOnDemandMeasure]
2022-12-03 14:55:45.980 1759-27462 SHW - Hear...Repository pid-1759 I [startOnDemandMeasure.awaitClose]
2022-12-03 14:55:45.980 1759-27462 SHW - Hear...Repository pid-1759 I [stopOnDemandMeasure]<<<
2022-12-03 14:55:45.989 1759-2138 SHW - HealthSensor pid-1759 I unregisterListener ~~listener 1 = 397d98d
```

> The result of this measurement being 73.

> further on I can see in the logs that the sensor has been updated with a value of 77. What is the reason for this difference?

```
2022-12-03 14:56:21.776 21221-21221 HRSensor io....stant.companion.android.debug D HR event received with accuracy: high and value: 77.0 with event count: 4
```

There is almost a 40 second gap in these logs it is totally reasonable to see an update of a higher value seconds later. This is what the sensor reports and it seems like in this case it took 4 attempts to get a valid reading. 

Looking at your logs I can see that even with 5 events we still get reports from the sensor saying no contact with a value of 0. I can try to increase this but I am not sure what more we can do asides from increasing this value. What we can see from the logs is that theory is correct we do indeed get a `no_contact` reading and then sometimes it becomes a valid reading. We rely on the sensor to give us a valid reading, we do not get any updates from any other apps. Let me try to increase the events to 10 to see if that improves the situation here. The good news is that the code is working and returning more values from before.

Please do not expect this value to be in sync with anything else on the watch.

How long did you let this code run for? Did you end up having long gaps like before? Here is my own graph showing gaps up to 2 hours in between readings which we consider to be normal.

![image](https://user-images.githubusercontent.com/1634145/205458158-3e7460c5-a591-4132-a379-caa688f65330.png)

**Comment by dshokouhi** [comment]
ok updated my PR to wait for 10 events <https://github.com/home-assistant/android/actions/runs/3609949216>

Please let this code run for sometime to see if the graph flattens like before.

**Comment by damiaorocha** [comment]
I have been using it for approximately 6 hours.

![image](https://user-images.githubusercontent.com/72225762/205459869-b2f453b0-facc-4326-b4b9-83625b9d85c9.png)

**Comment by dshokouhi** [comment]
ok lets see if this latest change improves the situation. By the way the 2 hour gaps we see here appear to be reasonable and on par with my own device.

**Comment by damiaorocha** [comment]
After 48 hours of using wear-debug, the heart rate sensor update seems to have behaved well
![Screenshot_20221205_181006_Home Assistant](https://user-images.githubusercontent.com/72225762/205743745-ed60fc0b-74ca-47c7-bbf6-ad05cc9d8276.jpg)

**Comment by dshokouhi** [comment]
Wow that does indeed look a lot better! Thank you for confirming that it is working with the latest changes. Honestly I did not think it would help but I am glad to be disproven here lol.

**Comment by damiaorocha** [comment]
Thanks

**Comment by barrelltitor** [comment]
Wow that is amazing! Hope this gets merged in the app soon!

**Comment by Pippisbruffi** [comment]
Yesterday I added my fossil gen 5 to home assistant and set a graph sensor on my main dashboard to display heartbeat frequency. This is the result: 

Problem is I still get heartbeat readings while I'm not wearing the watch... For example:
![image](https://user-images.githubusercontent.com/32282739/215315382-8958c5eb-210c-4d75-ac82-8d66c3e436f6.png)

Starting from 23:00 I was sleeping and the watch was charging on its stand, but as you can see the heartbeat sensor sent a lot of readings... Any idea why?

**Comment by porcupineofthemoon** [comment]
Hi all,
this problem seems still to be existing.
I have a Google Pixel Watch 2 with the latest Home Assistant App Version 2024.7.3.
Was the outcome of the experimenting from this thread included into the official version?

**Comment by dshokouhi** [comment]
if you guys still have an issue please file a new issue with logs so we can look into it. Personally speaking my sensor is updating as expected. This issue was fixed over a year ago.

![image](https://github.com/user-attachments/assets/aa58c6ec-bff4-40f4-9964-7b8c16a15914)
